include/mesh/distributed_mesh.h

```diff
@@ -81,7 +81,12 @@ class DistributedMesh : public UnstructuredMesh
    * Copy and move assignment are not allowed.
    */
   DistributedMesh & operator= (const DistributedMesh &) = delete;
-  DistributedMesh & operator= (DistributedMesh &&) = delete;
+
+  /**
+   * Overloaded operator= will move contents of other_mesh to calling
+   * MeshBase object.
+   */
+  virtual MeshBase & operator= (MeshBase && other_mesh) override;
```

Comment:
@jwpeterson I gather that move assignment not being virtual is the convention. A very typical use case for the assignment operator here will be along the lines,

system.get_mesh() = std::move(*mesh_two);

Now System::get_mesh() returns type MeshBase, whereas mesh_two can also be of type MeshBase, if for example it was declared to be a shared_ptr to an UnstructuredMesh. This is the case in the unit test I have added. 

With operator=(&&) being virtual, it appears that the proper derived class types for the system mesh and mesh_two (Replicated/Distributed in the unit test) were detected and the program gave the correct output.

If I remove the virtual qualifier, the program reverts to the MeshBase implementation rather than resolving the derived type of the calling object automatically.

Reply:
It looks like virtual assignment operators are ok if we ensure that base class move operators are only called once,
https://stackoverflow.com/questions/64325516/move-assignment-operator-and-virtual-inheritance/

Reply:
You're absolutely correct that the move assignment operator shouldn't be virtual ... but notice that it's kind of entirely our fault that he went down that path:

"I'm even more hesitant to create a precedent of having "named" copy or move assignment operators when the language provides a standard way of naming those functions."

As I should have realized at the time, the language *doesn't* provide a standard way of providing virtual move assignment, does it?  And virtual move assignment is what we need, to do a non-slicing move assignment into a target for which our only handle is a reference-to-base-class.  ```virtual assign(MeshBase &&)``` ought to be implemented as shims to ```ReplicatedMesh::operator=(ReplicatedMesh &&)``` and ```DistributedMesh::operator=(DistributedMesh &&)``` under the hood (after a libmesh_cast_ref to ensure that we're not trying to mix the two), but to do the move into a base class reference it's going to need to be virtual, so it ought to be a named assign() function itself, right?

Reply:
OK, damn, I failed to understand the use case which @vikramvgarg just articulated above, which clearly has to involve virtual functions of some kind or another, and not just standard move assignment. I think we are a lot closer than we were before, however. Your idea of a virtual assign that is overridden in the bases to call move assignment feels like a good approach to me, plus now we will also have normal move assignment for Meshes.